Magic Routing Hint

rajarshimaitra opened the pull request:
This code adds the Magic Routing Hint logic. As most of the logic is at the application layer, the code is added in the test cases of v2 swaps. 

No changes in the internals of the library is required.

This is untested. And to test it one has to perform the following.

- Initiate a reverse swap with Boltz.
- Use the invoice received in reverse swap and initiate a submarine swap with the same invoice.

rajarshimaitra committed:
update cargo toml

rajarshimaitra committed:
update boltz api to fetch mrh bip21

rajarshimaitra committed:
add magic routing hint routine

rajarshimaitra committed:
add magic routing hint code for bitcoin and liquid

michael1011 commented on src/util/mod.rs:
```diff
@@ -1,13 +1,18 @@
 pub mod ec;
 pub mod secrets;
 
+const ENDPOINT: &str = "https://api.testnet.boltz.exchange";
```
Not a fan of hardcoding that here. I don't think it is being used

rajarshimaitra replied:
Removed..

michael1011 commented on src/util/mod.rs:
```diff
@@ -1,13 +1,18 @@
 
+const ENDPOINT: &str = "https://api.testnet.boltz.exchange";
+const MAGIC_ROUTING_HINT_CONSTANT: &str = "0846c900051c0000";
+const LBTC_ASSET_HASH: &str = "144c654344aa716d6f3abcc1ca90e5641e4e2a7f633bc09fe3baf64585819a49";
```
The L-BTC asset hash depends on the network

rajarshimaitra replied:
I only found one asset hash from your doc. Added a placeholder for mainnet asset hash. Can be easily updated when I have it.

michael1011 commented on tests/bitcoin_v2.rs:
```diff
@@ -37,6 +38,36 @@ fn bitcoin_v2_submarine() {
+            .unwrap()
+            .require_network(bitcoin::Network::Testnet)
+            .unwrap();
+        let address_hash = sha256::Hash::hash(&Vec::from_hex(&bip21.address.to_string()).unwrap());
+        let msg = Message::from_digest_slice(address_hash.as_byte_array()).unwrap();
+
+        let receiver_sig = bitcoin::secp256k1::schnorr::Signature::from_slice(
+            &Vec::from_hex(&mrh_resp.signature).unwrap(),
+        )
+        .unwrap();
+
+        let receiver_pubkey = PublicKey::from_str(&route_hint.src_node_id.to_string())
+            .unwrap()
+            .inner;
+
+        secp.verify_schnorr(&receiver_sig, &msg, &receiver_pubkey.x_only_public_key().0)
+            .unwrap();
```
This shouldn't just be in a test but a public function that is available for consumers of the library to 
call
We might want to integrate this in the library function that creates a swap

rajarshimaitra replied:
There is no function as "create swap". Depends on the structure of the lib. Currently swap creating steps are in the test itself. Moved the MRH parts into it's own module `magic_routing.rs` and all the relevant functions are added there.

michael1011 commented on tests/bitcoin_v2.rs:
```diff
@@ -233,13 +262,19 @@ fn bitcoin_v2_reverse() {
     // Give a valid claim address or else funds will be lost.
     let claim_address = "tb1qq20a7gqewc0un9mxxlqyqwn7ut7zjrj9y3d0mu".to_string();
 
+    // Sign the address signature by claim priv key.
+    let address_hash = sha256::Hash::hash(&Vec::from_hex(&claim_address).unwrap());
+    let msg = Message::from_digest_slice(address_hash.as_byte_array()).unwrap();
+    let addrs_sig = Secp256k1::new().sign_schnorr(&msg, &our_keys);
```
Same here

rajarshimaitra replied:
Moved to magic_routing module.

michael1011 commented on tests/bitcoin_v2.rs:
```diff
@@ -233,13 +262,19 @@ fn bitcoin_v2_reverse() {
         from: "BTC".to_string(),
         to: "BTC".to_string(),
         preimage_hash: preimage.sha256,
+        address_signature: addrs_sig.to_string(),
```
You also need to pass the address itself

rajarshimaitra replied:
Done.

rajarshimaitra committed:
New MRH module.

Create a new Magic Routing Hint module with all the relevant code.
Remove them from utility module.

New place holder added for liquid mainnet asset hash.

Remove bip21 as uri parsing isn't supported for liquid. So manual URI string parsing is added.

rajarshimaitra committed:
Update the tests and boltz API.

rajarshimaitra committed:
test fixes

rajarshimaitra committed:
mrh fixes

rajarshimaitra committed:
update mrh handling

i5hi commented:
Thanks @rajarshimaitra !
